argo/lib/pipelines/cat/README.md

```diff
@@ -0,0 +1,53 @@
+# How the pipeline works
+
+## State management
+
+- volume that contains files_to_update.txt
+- db table `cat_ingestion`: 
+	- id
+	- argo_id
+	- last_scan
+	- num_files_to_update
+	- num_c4_files_updated
+	- num_metis_files_updates
+	- updated_at
+	- modified_at
+
+## Pipeline details
+
+### Job 1 - File discovery
+
+- checks cat_ingestion.last_scan in the db 
+- performs a regex search on the ftp server for all files that > last_run
```

Comment:
Seems a lot simpler this way than how the current airflow version works, but what if there was an error in the previous run to where some desired files could not be grabbed... seems to me like they'd be missed forever? that'd be a major flaw

We've had runs fail to grab files to metis due to the janus token expiring.  Or for c4, there could be a failure if the job runs during a c4 downage.  Or even momentary networking blip.

Reply:
good question - we can use some argo primitives to orchestrate it such that:

- any failures to any jobs stop the pipeline
- any future scheduled pipelines runs will be stopped until an error in a pipeline is fixed

Reply:
Seems like that would put the burden on a manual follow up to diagnose? And perhaps after a fixed token or restored c4 connection the path would safely be to just restart the pipeline?

The primary downside compared to the current system then just seems like if SOME files got pulled before the error, then they'd nolonger get skipped in the retry.  So the retry would involve some repeated work, but that seems okay.

And then I guess another downside to "any failures to any jobs stop the pipeline" would be that the c4-side doesn't necessarily need to be killed due to an error on the metis-side, or vice versa.  Maybe there can be a "last_scan" element for both c4 and metis, and its next value could be set when files are scanned, yet only get logged into the final "last_scan" spot at the end of the c4 / metis update jobs which actually pull the files?

Reply:
> Seems like that would put the burden on a manual follow up to diagnose? And perhaps after a fixed token or restored c4 connection the path would safely be to just restart the pipeline?

we can totally add retries and catch for certain error conditions before we manually halt the pipeline

Reply:
> So the retry would involve some repeated work, but that seems okay.

repeated work is okay - it depends on the job, sometimes we want to make the job and the data it processes atomic. For instance the the file discovery job should either fail or pass. 

whereas the uploader jobs will have the list of files and their md5's so its much easier to work in some redundancy checks here

Reply:
> And then I guess another downside to "any failures to any jobs stop the pipeline" would be that the c4-side doesn't necessarily need to be killed due to an error on the metis-side, or vice versa. 

This is a good point! Ill need to think about this...